Report logs only on failures in JUnit and ScalaTest

hubertp opened the pull request:
### Pull Request Description

This change simplifies the existing `MemoryAppender` in order to accommodate for setting up Loggers in both, JUnit and ScalaTest. This change adds support for reporting logs on failures via JUnit's `Rule`s.
Made sure that (in most cases) we use `JulHandler` as engine's log handler so that logs are appropriately forwarded to slf4j for processing.

The functionality stays as before. The only difference is that we now have logs being reported when a test fails.
Closes #13308.

### Important Notes

A `ReportLogsOnFailure` has already been used in LS but not in `runtime-integration-tests`. It also didn't work well with logs in engine. 

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      [TypeScript](https://github.com/enso-org/enso/blob/develop/docs/style-guide/typescript.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.

hubertp committed:
Report logs only on failures in JUnit and ScalaTest

This change simplifies the existing `MemoryAppender` in order to
accomodate for setting up Loggers in both, JUnit and ScalaTest.

A `ReportLogsOnFailure` has already been used in LS but not in
`runtime-integration-tests`. It also didn't work for logs in engine.
Junit implements "logs on failure" via rules.

hubertp committed:
Further transition of JUnit tests

More tests use the "report logs on failure only" feature.

hubertp committed:
Increase default log level for tests

`trace` level seems more useful now that logs are printed only when
necessary.

hubertp committed:
DRY when creating context with test log handler

hubertp committed:
More DRY

Rather than following all `newBuilder` by `withLogHandler` ones, making
the builder use `JulHandler` by default.
This change required changes to `ExecCompilerTest` and
`HelloWorldCacheTest` to make use of log messages rather than stdout.

hubertp committed:
Added appender rule to remaining tests

hubertp committed:
Use customn log handler by default

hubertp committed:
Move remaining Scala tests to reporting logs on failure

hubertp committed:
Drop `order` in Rule

So far we have not experienced ordering issues.

hubertp commented on build.sbt:
```diff
@@ -2710,6 +2710,7 @@ lazy val `runtime-integration-tests` =
         "-Dtck.inlineVerifierInstrument=false",
         "-Dpolyglot.engine.AllowExperimentalOptions=true"
       ),
+      Test / javacOptions += "-implicit:none",
```
Not related to this PR but reduces warnings noise.

hubertp committed:
More coverage

hubertp committed:
fmt

hubertp committed:
fix benchmarks that expect output in stdout

hubertp committed:
Merge branch 'develop' into wip/hubert/13308-logs-on-failure

JaroslavTulach commented on engine/runtime-integration-tests/src/test/java/org/enso/compiler/test/ExecCompilerTest.java:
```diff
@@ -528,7 +531,8 @@ public void illegalPrivateConversion() {
     var expectedErrMsg = DeclaredAsPrivate$.MODULE$.explain();
     var runMethod = module.invokeMember(Module.EVAL_EXPRESSION, "run");
     runMethod.execute(0);
-    assertThat(ctxRule.getOut(), containsString(expectedErrMsg));
```
- @Akirathan was a great fan of `assertThat`
- and I started to use it as well
- can we keep `assertThat` when already present?

JaroslavTulach commented on engine/runtime-integration-tests/src/test/java/org/enso/compiler/test/StaticAnalysisTest.java:
```diff
@@ -39,8 +40,8 @@ public abstract class StaticAnalysisTest {
-                  .option(RuntimeOptions.LOG_LEVEL, Level.INFO.getName())
-                  .option(RuntimeOptions.LOG_LEVEL, Level.SEVERE.getName())
+                  .option(RuntimeOptions.LOG_LEVEL, Level.FINE.getName())
+                  .logHandler(JulHandler.get())
```
Some tests are adding this `logHandler` and some don't: 
- How does one find out when to add it?
- Shouldn't it always be added by default?

hubertp replied:
It is, if you are using `ContextUtils` As tests that extend `StaticAnalysisTest` don't use that rule, this change is needed.

hubertp replied:
In this particular case, it wasn't necessary, indeed.

JaroslavTulach commented on lib/java/test-utils/src/main/java/org/enso/test/utils/ContextUtils.java:
```diff
@@ -408,12 +410,48 @@ private static Context.Builder defaultContextBuilder(String... permittedLanguage
+     * Shortcut for {@code withModifiedContext(b -> b.option(RuntimeOptions.LOG_LEVEL,
+     * logLevel.getName()).logHandler(logHandler))}.
+     */
+    public Builder withLogHandler(Level logLevel, Handler logHandler) {
```
- for a while I was asking whether one can call this method multiple times
- and whether the result is cumulative
- reading the docs answers my question

JaroslavTulach commented on lib/java/test-utils/src/main/java/org/enso/test/utils/ContextUtils.java:
```diff
@@ -408,12 +410,48 @@ private static Context.Builder defaultContextBuilder(String... permittedLanguage
 
+    /**
+     * Shortcut for {@code withModifiedContext(b -> b.option(RuntimeOptions.LOG_LEVEL,
+     * logLevel.getName()).logHandler(logHandler))}.
```
- Could this behavior be the default?
- And calling `withLogHandler(null, null)` or co. would be an opt-out?

Akirathan replied:
I second that. I would also like the default log handler the one that consumes everything and only prints it on failure.

hubertp replied:
This is the default already. See context builder. This method is when you **don't** want that (e.g. log handler should be stdout) as apparently some tests rely on that.

Akirathan commented on engine/runtime-integration-tests/src/test/java/org/enso/compiler/test/ExecStrictCompilerTest.java:
```diff
@@ -28,6 +30,8 @@ public class ExecStrictCompilerTest {
           .withModifiedContext(ctxBldr -> ctxBldr.option(RuntimeOptions.STRICT_ERRORS, "true"))
           .build();
 
+  @Rule public ReportLogsOnFailureRule appenderRule = new ReportLogsOnFailureRule();
```
Any test that already uses `@Rule ContextUtils` could implement the ReportLogsOnFailure logic by just setting a default LogHandler. Right? Why do we need yet another rule for such tests?

hubertp replied:
No. Setting `logHandler` is to ensure that engine logs are being passed to slf4j appender.

The extra rule ensures that a) every new test is run with a clean log b) on failure collected logs are flushed to ConsoleAppender. You don't want to just keep collecting logs and then print _all_ of that on the first failure as it would contain logs from tests that previously succeeded. We use `ContextUtils` rule _most_ of the time but not in all cases and that's problematic if we combine the two rules. I feel like having that separation makes things a bit more explicit.

Akirathan replied:
It could also be problematic that in many cases, we use `ContextUtils` as a `@ClassRule`, but `ReportLogsOnFailureRule` is always used as just `@Rule`. OK, let's keep them separated.

Akirathan reviewed (commented):
Having another `@Rule` definition seems like a duplicate on many tests that already have `@Rule ContextUtils`. Can't the new logging behavior be the default for `ContextUtils`?

hubertp commented:
I will have another run at all those changed tests and see if I can somehow eliminate the few exceptions that needed manual settings.

hubertp committed:
Merge branch 'develop' into wip/hubert/13308-logs-on-failure

hubertp committed:
Avoid concurrent modification errors

hubertp committed:
Address PR comments

Reduced redundant `logHandler` setup in various places.

hubertp committed:
Merge branch 'develop' into wip/hubert/13308-logs-on-failure

Akirathan reviewed (approved):
LGTM in general. Ideally, this behavior should somehow be enforced for all the tests. In a few months, when we all start forgetting to put `@Rule ReportLogsOnFailure` in our tests, we will be in the same situation and with the same polluted outputs on the CI. Were you able to investigate if it is possible to somehow hook into jUnit or ScalaTest runner?

hubertp commented:
> we will be in the same situation and with the same polluted outputs on the CI.

Quite the contrary. If this rule is not present, no logs would be printed.

> Were you able to investigate if it is possible to somehow hook into jUnit or ScalaTest runner?

_Hook into_? To do what exactly?

hubertp committed:
Merge branch 'develop' into wip/hubert/13308-logs-on-failure

Akirathan commented:
> Quite the contrary. If this rule is not present, no logs would be printed.

Can you elaborate? runtime-integration-tests/src/test/resources/application-test.conf specifies that the default appender is MemoryAppender which forwards to console. So if I implement a new test in runtime-integration-tests without any rules, the logs will still be forwarded to the console even if the test is successful, no? And what about tests in other projects? Like project-manager, language-server, etc?

> Hook into? To do what exactly?

To do the same thing that you have implemented in the new rule - capture all output and only print it if a test failed.

hubertp committed:
Fix benchmark runs

In non-strict mode errors are being reported as warnings. Should
:fingers-crossed: help with CI failures.

hubertp commented:
> Can you elaborate? runtime-integration-tests/src/test/resources/application-test.conf specifies that the default appender is MemoryAppender which forwards to console. So if I implement a new test in runtime-integration-tests without any rules, the logs will still be forwarded to the console even if the test is successful, no?

No. The forwarding is manually requested. If you want to always see the logs you can change the default appender to console.

> To do the same thing that you have implemented in the new rule - capture all output and only print it if a test failed.

Let's take it offline as I'm still missing the point.

hubertp commented:
> > Can you elaborate? runtime-integration-tests/src/test/resources/application-test.conf specifies that the default appender is MemoryAppender which forwards to console. So if I implement a new test in runtime-integration-tests without any rules, the logs will still be forwarded to the console even if the test is successful, no?
> 
> No. The forwarding is manually requested. If you want to always see the logs you can change the default appender to console.

Reading it again, I feel that what you are asking makes sense. I remember that the reason why I _didn't_ do it was because of some logs that _could_ be reported prior to execution of any test cases, in setup.

JaroslavTulach commented:
Outdated PR with conflicts. Closing.
